#160451501 Create and update user profile

oladimillion opened the pull request:
**What does this PR do?**

Users should be able to create their profile

**Description of Task to be completed?**

- add create user profile route
- add update user profile route
- add get user profile route


**What are the relevant pivotal tracker stories?**
Story type: feature
Story title: Users should be able to create their profile
Story ID: #160451501

#160451496

oladimillion committed:
feature (user-profile) add user profile
  - add profile model
  - add profile migration
  - add cloudinary package

oladimillion committed:
feature (user-profile) add comments
      - add js doc comments

oladimillion committed:
feature (user-profile) add comments and tests
      - add js doc comments
      - add user profile tests

udochukwu committed:
Merge pull request #8 from andela/feature/160451496/signup-jwt

oladimillion committed:
fix merge conflict

oladimillion committed:
feature (user-profile) modify some route
  - modify update profile route
  - modify get single profile detail route

oladimillion committed:
fix bug

oladimillion committed:
add authorization option to the request headers

kampkelly commented:
LGTM

s-akinrele commented on server/controllers/profileController.js:
```diff
@@ -0,0 +1,150 @@
+/* eslint-disable no-unused-vars */
```
please remove this.

s-akinrele commented on server/controllers/profileController.js:
```diff
@@ -0,0 +1,150 @@
+  * @param {object} res - The response payload sent back from the controller
+  * @returns {object} - status, message and profile detail
+  */
+  static get(req, res) {
```
change this method name.

s-akinrele commented on server/controllers/profileController.js:
```diff
@@ -0,0 +1,150 @@
+  * @param {object} res - The response payload sent back from the controller
+  * @returns {object} - status, message and profile detail
+  */
+  static getAll(req, res) {
```
Use a more descriptive method name

s-akinrele reviewed (changes requested):
Make those changes as discussed.

oladimillion commented on server/controllers/userController.js:
```diff
@@ -23,8 +23,7 @@ export default class UserController {
         .then((user) => {
-          const { id } = user;
-          const token = generateToken({ id });
+          const token = generateToken({ id: user.id, username: user.username, email: user.email });
```
I add `username, email, and userid` into the `generateToken` function. Extra request to get this information from the server is not needed since they can be easily retrieved from the token when decoded.

oladimillion commented on server/helpers/profile.js:
```diff
@@ -0,0 +1,17 @@
+
+const { Profile } = Model;
+
+export default (req, res, next) => Profile.findOne({
```
This is an helper function to check whether a profile exist or not.

oladimillion commented on server/middlewares/authenticator.js:
```diff
@@ -10,7 +10,8 @@ export default {
 
   checkToken(req, res, next) {
-    const token = req.body.token || req.headers['x-access-token'] || req.headers.bearer;
+    // add 'authorization' option to the request headers
```
I add `authorization` option to the request headers

oladimillion commented on server/middlewares/authenticator.js:
```diff
@@ -10,7 +10,8 @@ export default {
   checkToken(req, res, next) {
-    const token = req.body.token || req.headers['x-access-token'] || req.headers.bearer;
+    // add 'authorization' option to the request headers
+    const token = req.headers['x-access-token'] || req.headers.authorization;
```
I add `authorization` option to the request headers in addition to the `x-access-token`

oladimillion commented on server/models/user.js:
```diff
@@ -3,7 +3,8 @@ export default (sequelize, DataTypes) => {
       unique: true,
-      allowNull: false
+      allowNull: false,
+      primaryKey: true,
```
`Profile model` references `User model` via username. I therefore make `User model's username field` a  primary key so that Sequelize association between the two models can work effectively.

oladimillion committed:
implement feedback

oladimillion committed:
fix merge conflict

oladimillion commented on server/tests/user.test.js:
```diff
@@ -149,7 +149,7 @@ describe('Users', () => {
           expect(res.status).to.equal(201);
           expect(res.body).to.have.property('token');
-          expect(res.body.message).to.equal('You have successfully signed up!');
+          expect(res.body.message).to.equal('User successfully signed up');
```
I changed this because, server response does not match the test case, causing the test to fail.

oladimillion commented on server/migrations/20180927153526-create-user.js:
```diff
@@ -9,6 +10,7 @@ module.exports = {
     },
     username: {
       allowNull: false,
+      primaryKey: true,
```
`Profile model` references `User model` via username. I therefore make `User model's username field` a  primary key so that Sequelize association between the two models can work effectively.

oladimillion commented on server/migrations/20180927153526-create-user.js:
```diff
@@ -34,7 +36,8 @@ module.exports = {
+
   }),
-  down: queryInterface => queryInterface.dropTable('Users')
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users')
```
This happened when I was resolving conflict.

onwuvic commented on package.json:
```diff
@@ -8,6 +8,9 @@
+    "migrate-undo": "sequelize db:migrate:undo:all",
+    "migrate": "npm run migrate-undo && sequelize db:migrate",
+    "reset:db": "sequelize db:migrate:undo:all && sequelize db:migrate",
     "migrate-reset": "sequelize db:migrate:undo:all && sequelize db:migrate",
```
migrate-reset and reset:db are doing the same thing. I don't see the need for reset:db

oladimillion replied:
done

onwuvic commented on server/controllers/profileController.js:
```diff
@@ -0,0 +1,149 @@
+   * @param {object} res - The response payload sent back from the controller
+   * @returns {object} - status, message and profile detail
+   */
+  static getProfiles(req, res) {
```
Great job, but use a more descriptive method name. (e.g. getAllProfile and getOneProfile is preferable).

oladimillion replied:
Done

onwuvic reviewed (commented):
LGTM

kampkelly commented:
LGTM

udochukwu commented:
LGTM

udochukwu reviewed (commented):
LGTM

slimsolz reviewed (commented):
LGTM

oladimillion committed:
feature(create-user): rename some profile controller function

Files modify/create:
- modified server/controllers/profileControllers.js
- modified package.json

Finished #160451501

s-akinrele reviewed (approved):
GTM
